src/internal/operations/backup.go

```diff
@@ -154,11 +163,139 @@ func (op *BackupOperation) Run(ctx context.Context) (err error) {
 	return err
 }
 
+// calls kopia to retrieve prior backup manifests, metadata collections to supply backup heuristics.
+func produceManifestsAndMetadata(
+	ctx context.Context,
+	kw *kopia.Wrapper,
+	sw *store.Wrapper,
+	sel selectors.Selector,
+	acct account.Account,
+) ([]*snapshot.Manifest, []data.Collection, error) {
+	complete, closer := observe.MessageWithCompletion("Fetching backup heuristics:")
+	defer func() {
+		complete <- struct{}{}
+		close(complete)
+		closer()
+	}()
+
+	m365, err := acct.M365Config()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		tid         = m365.AzureTenantID
+		oc          = selectorToOwnersCats(sel)
+		ms          = kopia.FetchPrevSnapshotManifests(ctx, kw.Conn(), &oc)
+		collections []data.Collection
+	)
+
+	for _, man := range ms {
+		bup := backup.Backup{}
+		if err := sw.Get(
+			ctx,
+			model.BackupSchema,
+			model.StableID(man.Tags[kopia.BackupIDTag]),
+			&bup,
+		); err != nil {
+			return nil, nil, err
+		}
+
+		for _, fn := range []string{graph.DeltaTokenFileName} {
+			for ro := range oc.ResourceOwners {
+				for _, sc := range oc.ServiceCats {
+					colls, err := collectMetadata(ctx, kw, fn, tid, ro, bup.SnapshotID, sc)
+					if err != nil {
+						return nil, nil, err
+					}
+
+					collections = append(collections, colls...)
+				}
+			}
+		}
+	}
+
+	return ms, collections, err
+}
+
+func collectMetadata(
+	ctx context.Context,
+	kw *kopia.Wrapper,
+	fileName, tenantID, resourceOwner, snapshotID string,
+	sc kopia.ServiceCat,
+) ([]data.Collection, error) {
+	paths := []path.Path{}
+	pathsByRef := map[string][]string{}
+
+	p, err := path.Builder{}.
+		Append(fileName).
+		ToServiceCategoryMetadataPath(
+			tenantID,
+			resourceOwner,
+			sc.Service,
+			sc.Category,
+			true)
+	if err != nil {
+		return nil, errors.Wrap(err, "building metadata path")
+	}
+
+	dir, err := p.Dir()
+	if err != nil {
+		return nil, errors.Wrap(err, "malformed metadata path: "+p.String())
+	}
+
+	paths = append(paths, p)
+	pathsByRef[dir.ShortRef()] = append(pathsByRef[dir.ShortRef()], fileName)
+
+	dcs, err := kw.RestoreMultipleItems(ctx, snapshotID, paths, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "collecting prior metadata")
+	}
+
+	return dcs, nil
+}
+
+func selectorToOwnersCats(sel selectors.Selector) kopia.OwnersCats {
+	service := sel.PathService()
+	oc := kopia.OwnersCats{
+		ResourceOwners: map[string]struct{}{},
+		ServiceCats:    map[string]kopia.ServiceCat{},
+	}
+
+	ros, err := sel.ResourceOwners()
+	if err != nil {
+		return oc
+	}
+
+	for _, sl := range [][]string{ros.Includes, ros.Filters} {
+		for _, ro := range sl {
+			oc.ResourceOwners[ro] = struct{}{}
+		}
+	}
+
+	pcs, err := sel.PathCategories()
```

Comment:
what does it mean if `sel.PathCategories()` returns an error? I'm not sure if it's safe in all cases to tell KW to look up snapshots with just service/categories

Reply:
It means the selector service instance (ex: ExchangeBackupSelector) doesn't comply with the interface used by this function.  Not something we should ever expect to see an error for in the wild.

Reply:
even though it's annoying/less efficient to deal with, it would be better to either return the error or fallback to doing a full backup (implementor's choice)

Reply:
I'll defer to the latter, since I don't think it's valuable to return the error here.